packages/blitz-next/src/index-browser.tsx

```diff
@@ -137,7 +148,53 @@ const setupBlitzClient = <TPlugins extends readonly ClientPlugin<object>[]>({
   //   allMiddleware.push(middleware)
   // }
 
-  const exports = plugins.reduce((acc, plugin) => ({...plugin.exports(), ...acc}), {})
+  const {middleware, events, exports} = plugins.reduce(
+    (acc, plugin) => ({
+      middleware: {
+        beforeHttpRequest: plugin.middleware.beforeHttpRequest
+          ? pipe<RequestInit>(acc.middleware.beforeHttpRequest, plugin.middleware.beforeHttpRequest)
+          : acc.middleware.beforeHttpRequest,
+        beforeHttpResponse: plugin.middleware.beforeHttpResponse
+          ? pipe<Response>(acc.middleware.beforeHttpResponse, plugin.middleware.beforeHttpResponse)
+          : acc.middleware.beforeHttpResponse,
+      },
+      events: {
+        onRpcError: plugin.events.onRpcError
+          ? merge<Error, Promise<void>>([
+              ...(Array.isArray(acc.events.onRpcError)
+                ? acc.events.onRpcError
+                : [acc.events.onRpcError]),
+              plugin.events.onRpcError,
+            ])
+          : acc.events.onRpcError,
+        onSessionCreated: plugin.events.onSessionCreated
+          ? merge<void, Promise<void>>([
+              ...(Array.isArray(acc.events.onSessionCreated)
+                ? acc.events.onSessionCreated
+                : [acc.events.onSessionCreated]),
+              plugin.events.onSessionCreated,
+            ])
+          : acc.events.onSessionCreated,
+      },
+      exports: {...plugin.exports(), ...acc.exports},
+    }),
+    {
+      middleware: {
+        beforeHttpRequest: pipe<RequestInit>(),
+        beforeHttpResponse: pipe<Response>(),
+      },
+      events: {
+        onRpcError: merge<Error, Promise<void>>([]),
+        onSessionCreated: merge<void, Promise<void>>([]),
+      },
+      exports: {},
+    },
+  )
+
+  globalThis.__BLITZ_beforeHttpRequest = middleware.beforeHttpRequest
+  globalThis.__BLITZ_beforeHttpResponse = middleware.beforeHttpResponse
+  globalThis.__BLITZ_onRpcError = events.onRpcError
+  globalThis.__BLITZ_onSessionCreated = events.onSessionCreated
```

Comment:
this also needs to go in `blitz` package

Reply:
Moved this logic to `reduceBlitzPlugins` function in `blitz/index-browser.ts` that will perform the `plugin.reduce`, assign the global middleware functions and create the required event listeners